File: src/clj/principles_systems_architecture/chapter1/section3/money.clj
(ns principles-systems-architecture.chapter1.section3.money
  (:require [schema.core :as s])
  (:import (principles_systems_architecture.chapter1.section3.quantity Quantity)))

(s/defrecord Money
    [value :- Long])

(s/defn money :- Money
  [value :- Long]
  (->Money value))

(s/defn minus :- Money
  [this :- Money
   other :- Money]
  (money (- (:value this) (:value other))))

(s/defn multiply :- Money
  [this :- Money
   quantity :- Quantity]
  (money (* (:value this) (:value quantity))))

File: src/java/principles_systems_architecture/chapter1/section3/Money.java
package principles_systems_architecture.chapter1.section3;

class Money {
    final int value;

    Money(int value) {
        this.value = value;
    }

    Money minus(Money other) {
        return new Money(this.value - other.value);
    }

    Money multiply(Quantity quantity) {
        return new Money(this.value * quantity.value);
    }

    @Override
    public String toString() {
        return String.format("Money(value = %s)", this.value);
    }
}
